flowtorch/bijectors/compose.py

```diff
@@ -72,8 +77,23 @@ def log_abs_det_jacobian(
             torch.zeros_like(y),
             self.domain.event_dim,
         )
+
+        if isinstance(x, BijectiveTensor) and x.has_ancestor(y):
+            # If x is a BijectiveTensor and has y as ancestor, then the inversion flow.inverse(y) = x has already
+            # been computed and we can recover the chain of parents instead of re-computing it.
+            _use_cached = True
+            parents = []
+            while isinstance(x, BijectiveTensor) and x is not y:
+                parents.append(x)
+                x = x.parent
+        else:
+            _use_cached = False
+
         for bijector in reversed(self.bijectors):
```

Comment:
I think we should refactor this part of the method: remove lines 81-91, and do the `isinstance` inside the `for` loop

Reply:
The reason why I implemented this is the following.
There are 2 cases: either we have

```python
y = flow.forward(x)
flow.log_abs_det_jacobian(x, y)
```

In that case, `y` is a bijective tensor and for every layer of the flow, `inverse` will returned the cached value of `layer.inverse(y)`.

In the second case we have
```python
x = flow.inverse(y)
flow.log_abs_det_jacobian(x, y)
```
There it is more tricky: when we do the loop in the `log_abs_det_jacobian` method, we call `inverse` for every layer, but actually this was already done when calling `flow.inverse`. Unfortunately, we can't know that by looking at `y` (which you can assume to be a `torch.Tensor`). The only way to know that `x` if a function of `y` is to look for that before the loop IINW. 
Of course I might be missing something... If you have a suggestion, I'd be happy to read it!